### PR TITLE
WIP: hyperkernel mods

### PR DESCRIPTION
**Pull request by cjams**

### CODE REVIEW

File: bfvmm/src/hve/arch/intel_x64/vmexit/init_signal.cpp [hunk 1]
```diff
@@ -60,7 +60,7 @@ init_signal_handler::handle(gsl::not_null<vcpu_t *> vcpu)
         vmcs_n::guest_activity_state::wait_for_sipi
     );
 
-    return true;
+    return (m_vcpu->global_state()->init_called = true);
```
Comment (rianquinn) [inline bfvmm/src/hve/arch/intel_x64/vmexit/init_signal.cpp]: Why did you add this back? My plan is to actually remove this variable completely from the global state as it is no longer needed (it was the sync bit)

File: bfvmm/src/hve/arch/intel_x64/vmexit/interrupt_window.cpp [hunk 2]
```diff
@@ -41,8 +41,16 @@ interrupt_window_handler::interrupt_window_handler(
...
-    if (this->is_open() && m_interrupt_queue.empty()) {
-        this->inject_external_interrupt(vector);
+    if (this->is_open()) {
+
+        if (m_interrupt_queue.empty()) {
+            this->inject_external_interrupt(vector);
+            return;
+        }
+
+        this->inject_external_interrupt(m_interrupt_queue.pop());
```
Comment (rianquinn) [inline bfvmm/src/hve/arch/intel_x64/vmexit/interrupt_window.cpp:51]: What happens if you inject more than one interrupt? Would this cause the previously injected interrupt to be lost?

Also.... do we need to guard this queue?

**Review by rianquinn — CHANGES_REQUESTED** [review]

File: bfvmm/src/hve/arch/intel_x64/vmexit/interrupt_window.cpp (hunk 2 above)
Comment (cjams) [reply bfvmm/src/hve/arch/intel_x64/vmexit/interrupt_window.cpp:51]: Only one will be injected. The issue with what was here before is that injection was happening only.  if the queue was empty. But it's possible to have interrupts arrive while the window is closed.

The queue is local to each vcpu so no. If we had interrupts enabled in the vmm then yes

Comment (rianquinn) [reply bfvmm/src/hve/arch/intel_x64/vmexit/interrupt_window.cpp:51]: So that seems like an issue. If you queue up more than one and the vector is not the same, this code will drop the vector. Any ideas on how to fix this?

I also agree that the previous code was not right.
